feat(xo6/core): improve color context (#7398)
vatesfr/xen-orchestra: @xen-orchestra/web-core/docs/contexts/color-context.md

<!-- TOC -->

- [Overview](#overview)
  - [CSS variables](#css-variables)
  - [Available color contexts](#available-color-contexts)
  - [Usage](#usage)
  <!-- TOC -->

# Overview

The color context provides a way to apply a set of colors variants to a component and all its descendants at any depth.

Each descendant has the ability to change the context value, affecting itself and all of its descendants at any level.

The purpose is to colorize a component and its descendants by applying a single CSS class on the parent component (e.g., applying the class on a modal component container will style all children components using the context).

## CSS variables

The color context relies on the usage of the following variables:

```css
--color-context-primary;
--color-context-primary-hover;
--color-context-primary-active;
--color-context-primary-disabled;

--color-context-secondary;
--color-context-secondary-hover;

> --color-context-secondary-active;

Reviewer: Is it expected that we don't have a `--color-context-secondary-disabled` variable?

Author: Yes, I did not find any use case during my tests with the existing components. Maybe we can add it later if needed ?

When you set a color context, the variables are updated with the help of CSS classes defined in `_context.pcss`:

```css
.color-context-info {
  --color-context-primary: var(--color-purple-base);
  --color-context-primary-hover: var(--color-purple-d20);
  --color-context-primary-active: var(--color-purple-d40);
  --color-context-primary-disabled: var(--color-grey-400);

  --color-context-secondary: var(--background-color-purple-10);
  --color-context-secondary-hover: var(--background-color-purple-20);
  --color-context-secondary-active: var(--background-color-purple-30);
}

.color-context-success {
  --color-context-primary: var(--color-green-base);
  --color-context-primary-hover: var(--color-green-d20);
  /*...*/
}
```

You can add any other context by adding a `color-context-<my-context>` class and setting the desired values for the variables.

**Important note: remember to set a value for all variables to avoid any missing styles.**

## Available color contexts

Color contexts rely on the type `Color` defined in `/lib/types/color.type.ts`:

- `info` (_purple_)
- `success` (_green_)
- `warning` (_orange_)
- `error` (_red_)

## Usage

> To get and set the color context in a component, you can use the `useContext` composable, which takes a `ColorContext` (defined in `context.ts`) as first parameter and returns the `colorContextClass`:

Reviewer: ```suggestion
To get and set the color context in a component, you can pass the `ColorContext` to `useContext` and apply the `colorContextClass` to the root component:
```
